samples/s02-instances/hostCode.cpp

```diff
@@ -0,0 +1,342 @@
+  double xpos = 0.f, ypos = 0.f;
+  double lastxpos, lastypos;
+  do 
+  {
+    float speed = .001f;
+    lastxpos = xpos;
+    lastypos = ypos;
+    gprtGetCursorPos(context, &xpos, &ypos);
+    if (firstFrame) {
+      lastxpos = xpos;
+      lastypos = ypos;
+    }
+    int state = gprtGetMouseButton(context, GPRT_MOUSE_BUTTON_LEFT);
+
+    // If we click the mouse, we should rotate the camera
+    // Here, we implement some simple camera controls
+    if (state == GPRT_PRESS || firstFrame)
+    {
+      firstFrame = false;
+      float4 position = {lookFrom.x, lookFrom.y, lookFrom.z, 1.f};
+      float4 pivot = {lookAt.x, lookAt.y, lookAt.z, 1.0};
+      #ifndef M_PI
+      #define M_PI 3.1415926f
+      #endif
+
+      // step 1 : Calculate the amount of rotation given the mouse movement.
+      float deltaAngleX = (2 * M_PI / fbSize.x);
+      float deltaAngleY = (M_PI / fbSize.y);
+      float xAngle = (lastxpos - xpos) * deltaAngleX;
+      float yAngle = (lastypos - ypos) * deltaAngleY;
+
+      // step 2: Rotate the camera around the pivot point on the first axis.
+      float4x4 rotationMatrixX = rotation_matrix(rotation_quat(lookUp, xAngle));
+      position = (mul(rotationMatrixX, (position - pivot))) + pivot;
+
+      // step 3: Rotate the camera around the pivot point on the second axis.
+      float3 lookRight = cross(lookUp, normalize(pivot - position).xyz());
+      float4x4 rotationMatrixY = rotation_matrix(rotation_quat(lookRight, yAngle));
+      lookFrom = ((mul(rotationMatrixY, (position - pivot))) + pivot).xyz();
+
+      // ----------- compute variable values  ------------------
+      float3 camera_pos = lookFrom;
+      float3 camera_d00
+        = normalize(lookAt-lookFrom);
+      float aspect = float(fbSize.x) / float(fbSize.y);
+      float3 camera_ddu
+        = cosFovy * aspect * normalize(cross(camera_d00,lookUp));
+      float3 camera_ddv
+        = cosFovy * normalize(cross(camera_ddu,camera_d00));
+      camera_d00 -= 0.5f * camera_ddu;
+      camera_d00 -= 0.5f * camera_ddv;
+
+      // ----------- set variables  ----------------------------
+      gprtRayGenSet3fv    (rayGen,"camera.pos",   (float*)&camera_pos);
+      gprtRayGenSet3fv    (rayGen,"camera.dir_00",(float*)&camera_d00);
+      gprtRayGenSet3fv    (rayGen,"camera.dir_du",(float*)&camera_ddu);
+      gprtRayGenSet3fv    (rayGen,"camera.dir_dv",(float*)&camera_ddv);
+
+      gprtBuildShaderBindingTable(context, GPRT_SBT_RAYGEN);
+    }
+
+    // Calls the GPU raygen kernel function
+    gprtRayGenLaunch2D(context,rayGen,fbSize.x,fbSize.y);
+    
+    // If a window exists, presents the framebuffer here to that window
+    gprtBufferPresent(context, frameBuffer); 
+  }
+  // returns true if "X" pressed or if in "headless" mode
+  while (!gprtWindowShouldClose(context));
+
+  // Save final frame to an image
+  LOG("done with launch, writing frame buffer to " << outFileName);
+  gprtBufferSaveImage(frameBuffer, fbSize.x, fbSize.y, outFileName);
```

Comment:
Something funky is up with the image write. It looks like it's mirrored horizontally. Is it that way for you too?

Reply:
I noticed the image was flipped vertically. I thought I fixed this on master, maybe it wasn’t merged in? Is it also flipped horizontally though?

Reply:
That’s what I saw, yeah. I’ll send a screenshot of the window and the resulting image in a bit.

Reply:
Last night I thought about this for a bit. Could be a bug switching from BGRA to RGBA

Reply:
I might need to submit a PR on the swapchain branch to fix this.

Reply:
Interesting! Okay yea let's look into it. Gonna generate those images now...

Reply:
![Screenshot from 2022-12-13 10-35-11](https://user-images.githubusercontent.com/4563941/207391183-c926df96-949e-4dcc-b4dc-dedca14c44cc.png)
![s02-instances](https://user-images.githubusercontent.com/4563941/207391229-f058df5d-576a-43cb-a428-9dd4944e6c07.png)

Reply:
Looks like it's flipped in both in fact!

Reply:
This should now be fixed.

Reply:
What was the fix? (for posterity)

Reply:
In master I added a patch to the gprtBufferSaveImage call to swizzle the framebuffer before outputting it to a file. I also previously removed an stbi flip vertical call, just just didn’t have that merged into this branch before you ran this test.